src/python/pants/source/source_root.py

```diff
@@ -202,31 +222,59 @@ class OptionalSourceRoot:
 
 
 @rule
-def get_source_root(
+async def get_source_root(
     source_root_request: SourceRootRequest, source_root_config: SourceRootConfig
 ) -> OptionalSourceRoot:
     """Rule to request a SourceRoot that may not exist."""
-    matched_path = source_root_config.get_pattern_matcher().find_root(source_root_request.path)
-    if matched_path:
-        return OptionalSourceRoot(SourceRoot(path=str(matched_path)))
-    else:
-        return OptionalSourceRoot(None)
+    if ".." in str(source_root_request):
+        raise NoSourceRootError(
+            str(source_root_request), "`..` disallowed in source root searches. "
+        )
+
+    pattern_matcher = source_root_config.get_pattern_matcher()
+    # Ensure a relpath.
+    path = source_root_request.path
+    relpath = path.relative_to(_repo_root) if path.is_absolute() else path
```

Comment:
Would likewise be good to ban absolute paths in the constructor.

Reply:
Good idea, will do.